### DISCUSSION TITLE
tar_hook_before seems to break dependencies set via deps argument in tar_target_raw

### DISCUSSION THREAD

**Discussion by edalfon** [ropensci/tarchetypes#131] [Help]
Hi, 

We use a targets factory that internally sets dependencies using the `deps` argument from `tar_target_raw`. That works fine. But when we try to use `tar_hook_before` (or any tar_hook_*) on those targets, the pipeline does not recognize the dependencies anymore.

Here's a reprex, without tar_hook_before, where it works just fine.

```
targets::tar_dir({ 
  targets::tar_script({

    targets_factory_fn <- function() {

      upstream_targets <- list(
        targets::tar_target_raw("upstream1", "I am upstream 1 running"),
        targets::tar_target_raw("upstream2", "I am upstream 2 running")
      )

      downstream_targets <- targets::tar_target_raw(
        name = "downstream",
        command = "I am downstream running",
        deps = c("upstream1", "upstream2") # set the dependencies via `deps`
      )

      list(upstream_targets, downstream_targets)
    }


    targets <- list(
      targets_factory_fn()
    )

    targets
  })
  targets::tar_visnetwork(targets_only = TRUE)
})
```
![image](https://user-images.githubusercontent.com/8780149/211145632-46301858-026f-481b-8ae8-b8080c41edd7.png)

But the unexpected behavior arises when we use hooks. Here's the reprex, just like the working version above, but now using `tarchetypes::tar_hook_before` to prepend some code to the targets' command. It does not recognize the dependencies anymore.

```
targets::tar_dir({ 
  targets::tar_script({

    targets_factory_fn <- function() {

      upstream_targets <- list(
        targets::tar_target_raw("upstream1", "I am upstream 1 running"),
        targets::tar_target_raw("upstream2", "I am upstream 2 running")
      )

      downstream_targets <- targets::tar_target_raw(
        name = "downstream",
        command = "I am downstream running",
        deps = c("upstream1", "upstream2") # set the dependencies via `deps`
      )

      list(upstream_targets, downstream_targets)
    }


    targets <- list(
      targets_factory_fn()
    )

    tarchetypes::tar_hook_before(
      targets = targets,
      hook = print("Running hook.")
    )
  })
  targets::tar_visnetwork(targets_only = TRUE)
})
```

![image](https://user-images.githubusercontent.com/8780149/211145707-fd534325-d109-40c4-af17-8002ae0c900a.png)

And this happens also with `tar_hook_inner` and `tar_hook_outer`. The same behavior applies whenever the targets produced by the target factory are passed through the `tar_hook_*`, even if the hook does not apply to those targets (e.g. using `names = starts_with("x")`).

It affects only dependencies set via `deps` argument. `targets` still recognizes the dependencies in the code. For example, if the downstream target's code depends on one upstream target. Here's the reprex above, but now inducing one of the dependencies via de target's command in one case, and only via `deps` arg in the other.

```
targets::tar_dir({ 
  targets::tar_script({

    targets_factory_fn <- function() {

      upstream_targets <- list(
        targets::tar_target_raw("upstream1", "I am upstream 1 running"),
        targets::tar_target_raw("upstream2", "I am upstream 2 running")
      )

      downstream_targets <- targets::tar_target_raw(
        name = "downstream",
        command = quote(upstream1),
        deps = c("upstream1", "upstream2") # set the dependencies via `deps`
      )

      list(upstream_targets, downstream_targets)
    }


    targets <- list(
      targets_factory_fn()
    )

    tarchetypes::tar_hook_before(
      targets = targets,
      hook = print("Running hook.")
    )
  })
  targets::tar_visnetwork(targets_only = TRUE)
})
```

![image](https://user-images.githubusercontent.com/8780149/211158920-39dc2f0a-93d6-468b-98d0-16a939e31669.png)

Is this an expected behavior? Are we missing something here or doing something wrong? Because we certainly expected the dependencies set via `deps` to work exactly the same, with or without using hooks. Any workaround?

Thanks!

**Comment by wlandau** [comment]
You have convinced me to add a new `set_deps` argument to the hook functions. In your case, you would write `set_deps = FALSE` in `tar_hook_before()`. Please give accd77a902368c507b610aeda85e4e98ddc97dbe a try with `remotes::install_github("ropensci/tarchetypes")`.

As you know, in most cases a target's dependencies are not manually specified with the `deps` argument, but rather detected automatically using static code analysis. A hook changes the command of one or more targets, so any automatically-detected dependencies need to be refreshed. Usually, the function inserted as the hook becomes a new dependency. `set_deps` is `TRUE` by default for this reason and to ensure this version of `tarchetypes` is compatible with previous ones.

**Comment by edalfon** [comment]
Many thanks for taking a look at it and coming back with a fix so quickly.

I did try https://github.com/ropensci/tarchetypes/commit/accd77a902368c507b610aeda85e4e98ddc97dbe, but it still does not do the trick. 

I forked the repo and debugged a bit and I think I know what is going on.

This line in `tar_replace_command` that tries to keep the dependencies when `set_deps` is FALSE, does not do the trick, because this function always receives `target$command$deps` as an empty vector (`character(0)`) 
```
deps = if_any(set_deps, NULL, target$command$deps)
```

So debugging backwards, I found the dependencies are getting lost actually earlier, when `tar_hook_before` copy the targets. This line:
```
targets <- tar_copy_targets(targets)
```

And it seems to me the ultimate culprit is `tar_copy_target`, that copies almost everything into a new `tar_target_raw`, but never copy the dependencies (I mean, it copies the command's expression `command = target$command$expr`, and therefore later the static code analysis will catch any dependencies there, but it does not copy the `target$command$deps`). 

I think adding this line in `tar_copy_target` to propagate the deps would do the trick:

```
    deps = target$command$deps,
```

I did try it, installing a version with that small change (`remotes::install_github("edalfon/tarchetypes", ref = "targets-deps")`) in [this Posit Cloud project](https://posit.cloud/content/5193985) and it seems to work.

Having said that, and acknowledging `set_deps = FALSE` works just fine for our use case, and being super thankful for such a quick solution, I am having trouble imagining a use case for `set_deps = TRUE`. I mean, would it be easier to just always let the dependencies propagate? Just a thought!

Anyway, thanks a lot again and please let me know what you think.

**Reply by wlandau** [reply]
Thanks, I see what you mean. 9742190e2283b4763938c7dcf202d599828f7146 should fix it.

> Having said that, and acknowledging set_deps = FALSE works just fine for our use case, and being super thankful for such a quick solution, I am having trouble imagining a use case for set_deps = TRUE. I mean, would it be easier to just always let the dependencies propagate? Just a thought!

If the inserted hook calls a user-defined function, then changes to that function may change the output of the target. In other words, the hook introduces a new dependency that `targets` should track for most pipelines. This is why the dependencies should usually be rescanned.

**Reply by edalfon** [reply]
> Thanks, I see what you mean. [9742190](https://github.com/ropensci/tarchetypes/commit/9742190e2283b4763938c7dcf202d599828f7146) should fix it.

Awesome!, Thanks a lot!

> If the inserted hook calls a user-defined function, then changes to that function may change the output of the target. In other words, the hook introduces a new dependency that `targets` should track for most pipelines. This is why the dependencies should usually be rescanned.

Yeap. That part I get it. Sorry I did not explain myself clearly. The case that I cannot clearly see is: 
Once you have already set dependencies via `deps` (which is admittedly a very rare/infrequent use case), when would you be interested in dropping them?

wouldn't be better to just have the best of both worlds, and let the thing resolve the dependencies while also keeping those set via `deps`?

I don't!, just a bit of random thoughts here.

**Reply by wlandau** [reply]
I think I understand your last point now. After a target object is created, there is no record of whether `deps` was set manually. Keeping track of that would require adding a special field to the `command` object in target objects. Target objects are already complicated, and it is super important that the architecture stays as simple as possible. I avoid adding extra elements to accommodate edge cases this rare.

**Reply by edalfon** [reply]
Thanks for the clarifications. What I think it was not fully clear to me, is that manually setting dependencies with the `deps` argument is mutually exclusive with the automatic detection. I guess in my head `deps` was an **additional** way to specify dependencies, that would **add up** `deps` to those automatically detected via static code analysis. And that's why an optional argument (`set_deps`) to keep `deps` in the hook fns seemed unnecessary, because I thought you could have both ways simultaneously. But since both ways are mutually exclusive, I get it know. Many many thanks for your help here and for targets; such an awesome tool!
